### PR TITLE
Add bmp581 driver

### PR DESCRIPTION
**Pull request by talhaHavadar**
Hello all,

Here you can find the implementation of bmp581 driver with I2C for ZephyrRTOS. I am using this driver currently in one of my projects so I am able to test it if something goes wrong. 

Can you please review the changes and give feedback about things I need to change?

I also have question regarding bmp581_user.h Is there a way to open this file to user? Like they can include it in their application code and use some of the definitions in there? I am trying to find best practices to use custom ATTR for example.

Thanks in advance

### CODE REVIEW

File: drivers/sensor/bmp581/bmp581.c [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+#ifdef CONFIG_BMP581_TRIGGER
+							   .trigger_set = bmp581_trigger_set,
+#endif
+							   .attr_set = bmp581_attr_set};
+
+#ifdef CONFIG_BMP581_TRIGGER
+#define BMP581_CONFIG(i)                                                                           \
+	static const struct bmp581_config bmp581_config_##i = {                                    \
+		.i2c = DEVICE_DT_GET(DT_INST_BUS(i)),                                              \
+		.i2c_addr = DT_INST_REG_ADDR(i),                                                   \
```
Comment (MaureenHelm) [inline drivers/sensor/bmp581/bmp581.c:549]: Use `I2C_DT_SPEC_INST_GET`

File: drivers/sensor/bmp581/bmp581.h [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+	uint16_t i2c_addr;
+#ifdef CONFIG_BMP581_TRIGGER
+	struct gpio_dt_spec input;
+#endif
+};
+
+#ifdef CONFIG_BMP581_TRIGGER
+int bmp581_trigger_init(const struct device *dev);
+int bmp581_trigger_set(const struct device *dev, const struct sensor_trigger *trig,
+		       sensor_trigger_handler_t handler);
```
Comment (MaureenHelm) [inline drivers/sensor/bmp581/bmp581.h:347]: Trigger support doesn't appear to be implemented yet. Please remove all trigger references until this feature is actually imoplemented

File: dts/bindings/sensor/bosch,bmp581.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+description: |
+    The BMP581 is a Barometric pressure sensor. See more info at:
+    https://www.bosch-sensortec.com/products/environmental-sensors/pressure-sensors/bmp581/
+
+compatible: "bosch,bmp581"
+
+include: i2c-device.yaml
```
Comment (MaureenHelm) [inline dts/bindings/sensor/bosch,bmp581.yml:7]:
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

File: drivers/sensor/bmp581/bmp581.c [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+	}
+#endif
+#define BMP581_INIT(i)                                                                             \
+	static struct bmp581_data bmp581_data_##i;                                                 \
+	BMP581_CONFIG(i);                                                                          \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(i, bmp581_init, NULL, &bmp581_data_##i, &bmp581_config_##i,          \
```
Comment (MaureenHelm) [inline drivers/sensor/bmp581/bmp581.c:563]: Use `SENSOR_DEVICE_DT_INST_DEFINE`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add this sensor to `tests/drivers/build_all/sensor`

> I also have question regarding bmp581_user.h Is there a way to open this file to user? Like they can include it in their application code and use some of the definitions in there? I am trying to find best practices to use custom ATTR for example.

Public definitions can be placed in `include/zephyr/drivers/sensor`

File: drivers/sensor/bmp581/bmp581.h (hunk 2 above)
Comment (talhaHavadar) [reply drivers/sensor/bmp581/bmp581.h]: Indeed, I have just removed the not implemented parts from the source.

File: drivers/sensor/bmp581/Kconfig [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#
+# Copyright (c) 2022 Badgerd Technologies B.V and its affiliates
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig BMP581
+	bool "BMP581"
+	depends on I2C && GPIO
```
[the remark this answers is not in the document]
Comment (talhaHavadar) [reply drivers/sensor/bmp581/Kconfig]: Updated as suggested.

File: drivers/sensor/bmp581/bmp581.c [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+			  struct bmp581_data *drv);
+static int set_odr_config(const struct sensor_value *odr, struct bmp581_data *drv);
+static int soft_reset(struct bmp581_data *drv);
+static int set_iir_config(const struct sensor_value *iir, struct bmp581_data *drv);
+static int get_power_mode(enum bmp5_powermode *powermode, struct bmp581_data *drv);
+static int set_power_mode(enum bmp5_powermode powermode, struct bmp581_data *drv);
+
+int reg_read(uint8_t reg, uint8_t *data, uint16_t length, struct bmp581_data *drv)
+{
+	return i2c_burst_read(drv->i2c, drv->i2c_addr, reg, data, length);
```
[the remark this answers is not in the document]
Comment (talhaHavadar) [reply drivers/sensor/bmp581/bmp581.c]: Updated as suggested.

File: drivers/sensor/bmp581/bmp581.c (hunk 1 above)
Comment (talhaHavadar) [reply drivers/sensor/bmp581/bmp581.c]: Updated as suggested.

File: drivers/sensor/bmp581/bmp581.c (hunk 4 above)
Comment (talhaHavadar) [reply drivers/sensor/bmp581/bmp581.c:540]: Updated as suggested

File: drivers/sensor/bmp581/bmp581.c [hunk 7]
```diff
@@ -0,0 +1,563 @@
...
+
+	if (val == NULL) {
+		return BMP5_E_NULL_PTR;
+	}
+
+	switch (chan) {
+	case SENSOR_CHAN_ALTITUDE: {
+		// val2 must have P0 (typically pressure at sea level or ground)
+		// calculate altitude in meters
+		float altitude = 44330 * (1.0 - pow(drv->last_sample.pressure / val->val2, 0.1903));
```
[the remark this answers is not in the document]
Comment (talhaHavadar) [reply drivers/sensor/bmp581/bmp581.c]: I did update the calculation and tried to simplify it a bit more. I also added the link to whitepaper that has the formula.

**Review by MaureenHelm — COMMENTED** [review]
Please incorporate requested changes by amending your original commits rather than adding new fixup commits. You also need to address the failed compliance check.

**Comment by talhaHavadar** [comment]
> Please incorporate requested changes by amending your original commits rather than adding new fixup commits. You also need to address the failed compliance check.

Just updated the commits and also fixed the coding style issues hopefully. Let's see if it is all green or not.

**Comment by carlescufi** [comment]
@gmarull @avisconti could you take a look?

File: drivers/sensor/bmp581/bmp581.c [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+
+int reg_read(uint8_t reg, uint8_t *data, uint16_t length, struct bmp581_data *drv)
+{
+	return i2c_burst_read_dt(&drv->i2c, reg, data, length);
+}
+
+int reg_write(uint8_t reg, const uint8_t *data, uint16_t length, struct bmp581_data *drv)
+{
+	return i2c_burst_write_dt(&drv->i2c, reg, data, length);
+}
```
Comment (MaureenHelm) [inline drivers/sensor/bmp581/bmp581.c:46]: Does this sensor also support SPI? If not, I don't see value in this wrapper layer.

File: drivers/sensor/bmp581/bmp581.c [hunk 9]
```diff
@@ -0,0 +1,565 @@
...
+	int ret = -1;
+
+	/* Reset the chip id. */
+	drv->chip_id = 0;
+	memset(&drv->osr_odr_press_config, 0, sizeof(drv->osr_odr_press_config));
+	memset(&drv->last_sample, 0, sizeof(drv->last_sample));
+
+	drv->i2c = cfg->i2c;
+
+	drv->i2c_addr = cfg->i2c_addr;
```
Comment (MaureenHelm) [inline drivers/sensor/bmp581/bmp581.c:520]: Why is this necessary?

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 10]
```diff
@@ -483,8 +483,8 @@ test_i2c_mpr: mpr@18 {
 };
 
 test_i2c_dps310: dps310@48 {
-		compatible = "infineon,dps310";
-		reg = <0x48>;
+	compatible = "infineon,dps310";
+	reg = <0x48>;
```
Comment (MaureenHelm) [inline tests/drivers/build_all/sensor/i2c.dtsi:487]: This change is unrelated to the bmp581 driver. Please revert or fix in a separate commit

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> You also need to address the failed compliance check.

This still needs to be fixed.

File: drivers/sensor/bmp581/bmp581.c (hunk 8 above)
Comment (talhaHavadar) [reply drivers/sensor/bmp581/bmp581.c:46]: yep sensor also supports SPI. But for now I didnt have time to implement SPI so I put it this way to make it easy to add in future.

File: drivers/sensor/bmp581/bmp581.c (hunk 9 above)
Comment (talhaHavadar) [reply drivers/sensor/bmp581/bmp581.c:520]: It is probably legacy. Let me check in next commit
